# Notebook 1A 
## Using Python (and a Jupyter notebook) as a calculator
## Total Points: 6
#### PHYS 225 Intro to Computational Physics, Fall 2020


## Using Python and Jupyter notebooks as a calculator

In [1]:
3+2.5  # hit shift+enter to execute a cell

5.5

In [2]:
# compute powers with the double star operator **
two_cubed = 2**3 # hit shift+enter to execute
two_tothefourth = 2**4 # notice the output isn't printed when the value is assigned to a variable

In [3]:
print(two_cubed) # adding a print() function around the calculation make sure each output is printed
print(two_tothefourth)

8
16


In [4]:
print(2*3*4) # multiplication is a "star"
print(7/2)  # division is a forward slash

24
3.5


Examples of square roots and other roots:
$$\sqrt{2} = 2^{1/2}$$
$$\sqrt[3]{8} = 8^{1/3} = 2$$

In [5]:
print(2** (1/2))  # one way to do a square root is calculating a number to the 1/2 power
print(8** (1/3))  # cube root

1.4142135623730951
2.0


### Task 1 (1 point, autograded)
In the Python cell below, calculate $\frac{(3-4)^2-8\times17}{5}$ and keep the value in a variable named `result1`.  Also, print `result1`.

The order of operation in python is similar to regular math, so you need to add parentheses in the right places.

Delete the `raise NotImplementedError()` as part of your solution. The error is used to check for incomplete assignments.

In [6]:
"""Compute the numerical value for the above expression in Task 1 and save the value to a variable named result1.
"""
result1 = 0 # replace 0 with python math statements
### BEGIN SOLUTION
result1 = ((3-4)**2 - 8*17)/5
print(result1)
### END SOLUTION

-27.0


Your code should print `-27.0`. Check that it does.  
Run the test case below. If the test case returns an error, your answer is not correct. If the test case runs, but no error is produced then everything is okay.

In [7]:
"""Check that result1 = -27.0"""
assert abs(result1 - (-27.0))/abs(-27.0) < 1e-4 # autograded with tolerance
assert result1 == -27.0  # match exact result

### What happens if you fail an autograded test?

Suppose you had the same `assert` statements, but the wrong numerical value, such as: 

    result1 = 5 
    
Run the code cell below to see what happens.

You can also try alternative answers, such as

    result1 = -27.0001 
    
which will pass the first test (0.01% accuracy) but fail the second test.

In [7]:
# Set the wrong answer
result1 = 5

"""Check that result1 = -27.0"""
assert abs(result1 - (-27.0))/abs(-27.0) < 1e-4 # autograded with tolerance
assert result1 == -27.0  # match exact result


AssertionError: 

Notes about autograding:
* The autograding stops at the first error. Once the first error is fixed, it is still possible to generate an error on a later assert statement.
* Because the cell above is not designated as an **answer cell**, failing the test does not penalize any points. Typically, I will only put assert statements within an **autograded answer cell**.

### Task 2  (1 point, autograded)
In the Python cell below, calculate $\sqrt[3]{3.4} - \frac{\sqrt{2.7 + 3.54}}{5}$ and keep the value in a variable named `result2`.  Also, print `result2`.

Remember to delete the `raise NotImplementedError()` as part of your solution. The error is used to check for incomplete assignments.

In [8]:
"""Compute the numerical value for the above expression in Task 2 and save the value to a variable named result2.
"""
result2 = 0 # replace 0 with python math statements
### BEGIN SOLUTION
result2 = (3.4)**(1/3) - (2.7 + 3.54)**(1/2)/5
print(result2)
### END SOLUTION

1.004094756333103


Your code should print `1.004094756333103` or something very close to that. Check that it does. Run the code below and make sure it does not produce an error.

In [9]:
"""Check that result2 = -1.004094756333103"""
assert abs(result2 - (1.004094756333103))/abs(1.004094756333103) < 1e-5 # autograded with tolerance

### Task 3 Example Using Python as a calculator for physics problems
Consider this totally typical UP1 problem (copied from the 2016 UP1 activity manual)

1. A ball is thrown upwards at 9 m/s toward a platform located 4.1 m above the point where you release the ball. Air resistance is negligable. *Find the time taken to reach the platform.*

Solution: Use displacement as a function of time 
$$y = y_0 + v_{y0}t + \frac{1}{2}at^2$$ 
and solve using the quadratic equation.
$$ t = \frac{-v_{y0}\pm \sqrt{v_{y0}^2 - 4(a/2)(y_0-y)}}{2(a/2)} $$

In [10]:
# Method 1 - plugging numbers directly into the quadratic equation
t1 = ((-9) + (9**2 - 4*(-9.8/2)*(0-4.1))**0.5)/(2*(-9.8/2)) # one root of quadratic equation
t2 = ((-9) - (9**2 - 4*(-9.8/2)*(0-4.1))**0.5)/(2*(-9.8/2)) # the other root
print( t1 ) # one root
print( t2 ) # the other root

0.8367346938775508
1.0


In [11]:
# Method 2 - defining values of each variable and using a "symbolic" equation for the formula
y0 = 0  # initial height
y = 4.1  # final height of the ball in meters
vy0 = 9  # initial upward velocity in m/s
a = -9.8  # acceleration due to gravity in m/s**2

t1 = (-vy0 + ( vy0**2 - 4*(0.5*a)*(y0-y) )**0.5)/( 2*0.5*a )
t2 = (-vy0 - ( vy0**2 - 4*(0.5*a)*(y0-y) )**0.5)/( 2*0.5*a )
print( t1 )
print( t2 )

0.8367346938775508
1.0


### Task 3 (1 point, manually graded)
Although both approaches  produce the same numerical result, Dr. Zwickl strongly encourages the second method where all the values are defined and a "symbolic" equation for the result is used.  

In the cell below, describe some advantages to the second method preferred by Dr. Zwickl.

In this cell, explain why the second method is preferrable in this cell. Double click on the Markdown cell to edit it. `Shift+Enter` will make it look nicely formatted.

## Tasks 4, 5 and and 6 - Genesis Crash 
![NASA model of spacecraft](GenSC_collection2_150dpi.jpg)
![Crash site](Genesis_crash_site_scenery.jpg)
Based on Young and Freedman *University Physics* 13th edition Exercise 5.13.

On September 8, 2004, the *Genesis* spacecraft crashed in the Utah desert because its parachute did not open. The 210-kg capsule hit the ground at 311 km/h and penetrated the soil to a depth of 81.0 cm.  
a. Assuming it to be constant, what was its acceleration (in g's) during the crash?  
b. What force did the ground exert on the capsule during the crash? Express this force in Newtons.  
c. For how long did this force last?

Use the second approach.
Do all unit conversions in python.

**Assume $g=9.8 \text{ m/s}^2$ **

### Task 4 - Describe your solution strategy (1 point, manual graded)
Describe process as a series of steps. Create an ordered list by typing markdown like this. It will auto number the entries.
```
1. First step
1. next step
1. next step
```

Type your strategy here.

### Task 5 - Calculate the acceleration during the impact (1 point, autograded)
Calculate the acceleration and set the result to a variable named `acc_gen`. Print the value of `acc_gen`.  
Remember to delete the `raise NotImplementedError()` as part of your solution. The error is used to check for incomplete assignments.

In [12]:
"""Compute the numerical value for the acceleration of the Genesis spacecraft (in g's) 
and save the value to a variable named acc_gen
"""

# start off by defining all the known variables
# replace 0 with python equation for acc_gen in terms of known variables
acc_gen = 0 

### BEGIN SOLUTION
v0 = -311 * 1000 / 3600 # 311 km/hr downward converted to m/s
v = 0 # spacecraft ends at rest
d0 = 0 # ground level initially
d = -0.81 # penetrates 81 cm into ground.
g = 9.8 # acceleration due to gravity
m = 210 # mass of spacecraft in kg

acc_gen = (v**2 - v0**2)/(2*(d-d0)) /g
print(acc_gen)
### END SOLUTION

470.0831521451744


The result should be an acceleration of about 470 g's. Check that is what you also got.

In [13]:
"""Check that acc_gen = 470.0831521451744"""
assert abs(acc_gen - (470.0831521451744))/abs(470.0831521451744) < 1e-5 # autograded with tolerance

### Compute the force during the collision (example)
Calculate the force of the ground on the spacecraft (in Newtons) and set the result to a variable named `force`.

$$ \vec{F}_{\textrm{net}} = \vec{F}_{\textrm{ground}} + \vec{F}_{\textrm{grav}} = m\vec{a} $$
$$ F_{\textrm{ground}} = ma + mg = m(a+g) $$

In [14]:
"""
quantities defined in another cell previously are in the kernel's memory 
and can be used in
"""
a = acc_gen * g  # convert to m/s^2
F_ground = m*a + m*g  # calculate force of ground on spacecraft
print(F_ground)  # force in Newtons  

969489.127114769


### Task 6, Calculate the duration of the collision force (1 point, autograded)
Use the previously defined quantities to calculate the time the force acted during the *Genesis* spacecraft's impact. Save the result in a variable named `t_impact`.

In [15]:
"""Compute the time the force acted during the Genesis spacecraft's impact
and save the value to a variable named t_impact
print(t)
"""
### BEGIN SOLUTION
t_impact = (v - v0)/a  
print(t_impact)
### END SOLUTION

0.018752411575562705


In [16]:
"""Check that t_impact = 0.018752411575562705"""
assert abs(t_impact - (0.018752411575562705))/abs(0.018752411575562705) < 1e-5 # autograded with tolerance